rethinkdb/datadog_checks/rethinkdb/document_db/query.py

```diff
@@ -0,0 +1,132 @@
+        results, such as those returned by a GROUP BY operation. One copy of the metric will be submitted for each
+        key/value item in the mapping.
+        Keys:
+        * `path` (required): dotted path to the mapping of interest in a `document`.
+        * `key_tag` (required): keys of the mapping will be submitted as this tag. Example: `'country'`.
+        * `value_type` (required): metric type of values in the mapping. Example: `'gauge'`.
+    """
+
+    def __init__(
+        self,
+        source,  # type: Callable[..., Iterable[Tuple[Any, List[str]]]]
+        name,  # type: str
+        prefix,  # type: str
+        metrics=None,  # type: List[MetricSpec]
+        enumerations=None,  # type: List[Enumeration]
+        groups=None,  # type: List[Group]
+    ):
+        self.source = source
+        self.name = name
+        self.prefix = prefix
+        self.metrics = [] if metrics is None else metrics
+        self.enumerations = [] if enumerations is None else enumerations
+        self.groups = [] if groups is None else groups
+
+    def _make_metric_from_spec(self, document, spec, tags):
+        # type: (Any, MetricSpec, List[str]) -> Metric
+        path = spec['path']
+        name = spec.get('name', path)
+        value = lookup_dotted(document, path=path)
+
+        if 'transformer' in spec and spec['transformer'] is not None:
+            value = spec['transformer'](value)
+
+        if not isinstance(value, (int, float)):  # pragma: no cover
+            raise RuntimeError('Expected float or int, got {!r} of type {}', value, type(value))
+
+        name = dotted_join((self.prefix, name))
+
+        return {'type': spec['type'], 'name': name, 'value': value, 'tags': tags}
+
+    def _make_metrics_from_enumeration(self, document, enumeration, tags):
+        # type: (Any, Enumeration, List[str]) -> Iterator[Metric]
+        values = lookup_dotted(document, path=enumeration['path'])  # type: Sequence
+
+        for index, value in enumerate(values):
+            item_tags = tags + ['{}:{}'.format(enumeration['index_tag'], index)]
+
+            for spec in enumeration['metrics']:
+                spec = spec.copy()
+                spec['name'] = dotted_join((enumeration['path'], spec['path']))
+                yield self._make_metric_from_spec(value, spec, tags=item_tags)
+
+    def _make_metrics_from_group(self, document, group, tags):
+        # type: (Any, Group, List[str]) -> Iterator[Metric]
+        mapping = lookup_dotted(document, path=group['path'])  # type: Mapping
+
+        for key in mapping:
+            item_tags = tags + ['{}:{}'.format(group['key_tag'], key)]
+            spec = {
+                'type': group['type'],
+                'name': group['path'],
+                'path': key,
+            }  # type: MetricSpec
+            yield self._make_metric_from_spec(mapping, spec, tags=item_tags)
+
+    def run(self, check=None, **kwargs):
```

Comment:
```suggestion
    def run(self, check, **kwargs):
```

Any reason check could be None?

Reply:
Not *have* to pass an `AgentCheck` just for the logger — see unit tests and example scripts? 😇

Reply:
2): Sounds good.

Reply:
You can probably just always pass the a `logger` instead of the `check` and in test pass a dummy logger. WDYT?

Reply:
Fixed in b9d3e92184424c188b06db6d4a104ed2b963715a :+1: